### ISSUE TITLE
New Python Api for the next pycf3 version

### ISSUE THREAD

**Issue by leliel12** [quatrope/pycf3#7]
Currently, the pycf3 API is defined in terms of searches given if the
the system used is *equatorial*, *galactic* o *super-galactic*

```python

cf3 = pycf3.CF3()

cf3.equatorial_search(ra, dec, distance=None, velocity=None)
cf3.galactic_search(glan, glon, distance=None, velocity=None)
cf3.supergalactic_search(gls, sgb, distance=None, velocity=None)

```

Internally this makes the validation quite simple,
we just have to validate that the user keeps in `None` or the *distance* or
*velocity*.

## Problems

- The `_search` verb of each method is not representative of the operation.
- The coordinate system is given too much prominence when what matters is whether the estimate is speed or distance.
  
  
## Proposals

We propose a new Python-API based on distance vs speed estimation. 

The coordinate system will be determined if the user supplies
the combination of one of the possible coordinates.

- `ra`, `dec` for *equatorial*.
- `glan`, `glon` for *galactic*.
- `slg`, `sgb` for *supergalactic*

Based on the following documentation from the cf3 page that says

> It computes expectation distances or velocities based on smoothed velocity 
  field from the Wiener filter model of Graziani et al. 2019.

The following alternatives are proposed:

```python
# Retrieve
cf3.retrieve_distance(velocity, ra, dec)
cf3.retrieve_velocity(distance, ra, dec)

# Expected
cf3.expected_distance(velocity, ra, dec)
cf3.expected_velocity(distance, ra, dec)

# Compute
cf3.compute_distance(velocity, ra, dec)
cf3.compute_velocity(distance, ra, dec)

# Get
cf3.get_distance(velocity, ra, dec)
cf3.get_velocity(distance, ra, dec)

```

IMHO `expected` or `retrieve` are the more correct. `compute` is too generic and 
`get` has another semantic in the underlying *requests* library.

**Comment by ekourkchi** [comment]
I think "calculate" or "compute" are better terms, since the original paper introduces the tool as a calculator.

So, the generic format of the calculate functions would be as follows

```
cf3 = pycf3.CF3()
cf3.caluclate(ra=None, dec=None, glat=None, glon=None, sgl=None, sgb=None, distance=None, velocity=None)

nam = pycf3.NAM()
nam.caluclate(ra=None, dec=None, glat=None, glon=None, sgl=None, sgb=None, distance=None, velocity=None)
```

These combinations of coordinates are acceptable:
 
- **ra** and **dec**
- **glat** and **glon**
- **sgl** and **sgb**

In a function call, users input either **velocity** or **distance** and not both. So, we make sure only one of these values are legitimate (i.e. not None)

Other constraints are:
- **sgb**, **glat** and **dec** range between -90 and 90 (inclusive)
- for NAM: the input distance < 38 (less than 38 Mpc)
- for CF3: the input distance < 200 (less than 200 Mpc)

### Example:

- cf3.caluclate(ra=194.2, dec=35, distance=20)
- cf3.caluclate(glat=60, glon=-23, velocity=1000)
- nam.calculate(sgl=225, sgb=18, distance=15)

**Comment by ekourkchi** [comment]
We should expect astronomers to be lazy with functions, meaning that we only provide them with a single **calculate** function and under the hood that functions know how to manage the API calls based on the inputs users pass.

**Comment by BrunoSanchez** [comment]
I think it may be possible to offer a wrapper global `.calculate` method with the arguments that you propose, but internally split the functionality as two separate methods depending on velocity or distance.
It would not hurt to have two separated methods, and if the user wants maybe offer a unified interface that just calls previously mentioned ones.

Just phrasing some opinions we gathered from users (shoutout to @erikpeterson23): Having two methods would make completely clear what the code does, and what the signature of the method requires. It is not trivial the exchange of distance and velocity, and two split methods would be easier to mantain, test and more importantly perhaps, to document.

**Comment by martinberoiz** [comment]
Hi all,

My first choice was to use the verb "estimate" but like Ehsan says, if it was introduced as a calculator, then maybe `calculate` is the reasonable choice here.

I agree in general with the proposal, one thing I'm not convinced though is the proliferation of arguments if we use just one function to query the distance and velocity.

As I see it, the combinatorial explosion of simultaneously dealing with the validation of mutually exclusive coordinate systems could create a lot of spaghetti code hard to maintain in the long run. But OTOH I do like the simple examples given. Query calls should be short and to the point. For example validating invalid things like cf3.calculate(ra=194.2, distance=20, velocity=100, glon=-23, sgl=225), invalid queries are just too many.

In my opinion, we should try to curb and restrict the valid queries using the API itself, so that there's no way for the client to request bad queries. Or at least make it easier to make the correct queries.

I apologize that I haven't thought of any specific example yet. I'm still mulling it over.

**Comment by ekourkchi** [comment]
Thanks a lot for your great effort and enthusiasm to develop the client interface.

For the sake of clarity we can separate methods for distance and velocity calculators. I agree with the potential of adding a lot of spaghetti code (which is not a good practice) to check for all various combinations.

The back-end API takes care of most errors and exceptions and returns a **message** with the output dictionary, informing users what might have gone wrong. If you notice anything out of ordinary, please let me know, I'd be happy to modify the API codes.

Concerning the client method calls, I am leaning to take coordinates first and then the type of coordinate system for clarity, as opposed to have three different methods for various coordinate systems. This adds more flexibility, because if we later add an **ecliptic** coordinate system to the API on the server side, then we don't need to add any additional method in the client side as well.

This can be the method templates:

```

    cf3.calculate_velocity(194.2, 35, system='equatorial', distance=20)
    cf3.calculate_distance(60, 23, system='galactic', velocity=1000)
    nam.calculate_velocity(225, 18, system='supergalactic', distance=15)


```
The default of the coordinate system can be set to **equatorial**.

Also, I have other ideas like returning the blue Distance-Velocity data that produces the blue curves in the Bokeh plots. The client side can benefit from those data and generate a similar plot, etc.

**Comment by leliel12** [comment]
One iteration more:

What everyone thinks about 

```python
cf3.calculate_velocity(ra=194.2, dec=35,  distance=20)
cf3.calculate_distance(glat=60, glon=23, velocity=1000)
nam.calculate_velocity(sgl=225, sgb=18, distance=15)

# ERROR: mix of systems
nam.calculate_velocity(ra=194.2, sgb=18, distance=15)

```

This version avoids the use of string parameters

**Comment by ekourkchi** [comment]
This is even better than my suggestion. I am totally happy with your recommendation.

**Comment by martinberoiz** [comment]
I liked Ehsan's proposal for the `system` argument, but I don't know what are the cons of using string parameters.

This is just a thought, but what about accepting an astropy SkyCoord object to remove ambiguity? Then we transform to whatever system.

**Comment by leliel12** [comment]
mmmm one iteration more:

```python
cf3.calculate_velocity(ra=194.2, dec=35,  distance=20)

# and also
c = SkyCoord(10, 20, unit="deg")  # defaults to ICRS frame
cf3.calculate_velocity(c,  distance=20)

# and also
c = SkyCoord([1, 2, 3], [-30, 45, 8], frame="icrs", unit="deg")  # 3 coords
cf3.calculate_velocity(c,  distance=20) # this came with 3 results
```

**Comment by leliel12** [comment]
and also the plus of astropy afiliated :D

**Comment by ekourkchi** [comment]
This is great. 

Is it hard to keep the original version as well?
I mean, if someone doesn't know how to use astropy (surprisingly, many fresh astronomy student don't know it yet). It's a plus if we keep the original functionalities and add the new ones.

**Comment by leliel12** [comment]
Yes the two version is the idea

But dont be afraid about the people that dont know about astropy, later or
soon they need to learn

I think the "search" api will be maintained for some time (until cf4?) But
will raise a deprecation warning when someone use it.

**Comment by ekourkchi** [comment]
The Cosmicflows-3 API would be maintained for quite a while. It takes at least another year to analyze the Cosmicflows-4 data and upgrade the calculator. Even after that, I anticipate we will still keep CF3 availble and running for longer period of time to enable researchers switch back and forth between two versions of the CF analyses.

**Comment by leliel12** [comment]
This is the implemented test that defines the entire behavior of the new cf3 API. 

We can take 3 days (or 3 thumbs up) to check if all is ok

Please click the smiley and select the "thumbs-up" to approve this, or add comments bellow

----

Test implemented


## `calculate_distance` tests

- `test_equatorial_calculate_distance_vel_EQ_10`:
  Validates the correct response of call *calculate_distance* with equatorial
  coordinates (`ra` and `dec`) and `velocity=10`.

- `test_galactic_calculate_distance_vel_EQ_10`:
  Validates the correct response of call *calculate_distance* with galactic
  coordinates (`glon` and `glat`) and `velocity=10`.

- `test_sgalactic_calculate_distance_vel_EQ_10`:
  Validates the correct response of call *calculate_distance* with supergalactic
  coordinates (`gsgl` and `sgb`) and `velocity=10`.


## `calculate_velocity` tests

- `test_equatorial_calculate_velocity_dis_EQ_10`:
  Validates the correct response of call *calculate_velocity* with equatorial
  coordinates (`ra` and `dec`) and `distance=10`.

- `test_galactic_calculate_velocity_dis_EQ_10`:
  Validates the correct response of call *calculate_velocity* with galactic
  coordinates (`glon` and `glat`) and `distance=10`.

- `test_sgalactic_calculate_velocity_dis_EQ_10`:
  Validates the correct response of call *calculate_velocity* with supergalactic
  coordinates (`gsgl` and `sgb`) and `distance=10`.


## Velocity or distance == 0 tests

- `test_calculate_distance_velocity_eq_0`: call the method
  *calculate_distance* with the `velocity=0` raises a `ValueError`.

- `test_calculate_velocity_distance_eq_0`: call the method
  *calculate_velocity* with the `distance=0` raises a `ValueError`.


## Velocity or distance < 0 tests

- `test_calculate_velocity_distance_lt_0`: call the method
  *calculate_velocity* with the `distance < 0` raises a `ValueError`.

- `test_calculate_distance_velocity_lt_0`: call the method
  *calculate_distance* with the `velocity < 0` raises a `ValueError`.


## alpha or delta not a number tests

- `test_calculate_velocity_alpha_not_number`: call the method
  *calculate_velocity* with `ra`, `glon` or `sgl` equals to something different
  than a integer or float raises a `TypeError`.

- `test_calculate_velocity_delta_not_number`: call the method
  *calculate_velocity* with `dec`, `glat` or `sgb` equals to something different
  than a integer or float raises a `TypeError`.

- `test_calculate_distance_alpha_not_number`: call the method
  *calculate_distance* with `ra`, `glon` or `sgl` equals to something different
  than a integer or float raises a `TypeError`.

- `test_calculate_distance_delta_not_number`: call the method
  *calculate_distance* with `dec`, `glat` or `sgb` equals to something different
  than a integer or float raises a `TypeError`.


## velocity or distance not a number tests

- `test_calculate_velocity_distance_not_number`: call the method
  *calculate_velocity* with `distance` equals to something different
  than a integer or float raises a `TypeError`.

- `test_calculate_distance_velocity_not_number`: call the method
  *calculate_distance* with `velocity` equals to something different
  than a integer or float raises a `TypeError`.


## Alpha > 90 or alpha < -90 tests

- `test_calculate_velocity_alpha_gt_90`: call the method
  *calculate_velocity* with `ra, glon sgl > 90` raises a `ValueError`.

- `test_calculate_velocity_alpha_lt_m90`: call the method
  *calculate_velocity* with `ra, glon sgl < -90` raises a `ValueError`.

- `test_calculate_distance_alpha_gt_90`: call the method
  *calculate_distance* with `ra, glon sgl > 90` raises a `ValueError`.

- `test_calculate_distance_alpha_lt_m90`: call the method
  *calculate_distance* with `ra, glon sgl < -90` raises a `ValueError`.

## Mix coordinates tests

- `test_calculate_velocity_mix_coordinate_system`: Call the method
  *calculate_velocity* with mixed coordinate sistem (using `ra` and `glon`
  for example) Raises a `pycf3.MixedCoordinateSystemError`.

- `test_calculate_distance_mix_coordinate_system`: Call the method
  *calculate_distance* with mixed coordinate sistem (using `ra` and `glon`
  for example) Raises a `pycf3.MixedCoordinateSystemError`.

- `test_calculate_velocity_multiple_alpha`: call the method *calculate_velocity*
  with multiple alpha (`ra`, `glat` for example) raises a `pycf3.MixedCoordinateSystemError`.

- `test_calculate_velocity_multiple_delta`:
  call the method *calculate_velocity*
  with multiple delta (`dec`, `glat` for example) raises a `pycf3.MixedCoordinateSystemError`.

- `test_calculate_distance_multiple_alpha`:
  call the method *calculate_distance*
  with multiple alpha (`ra`, `glat` for example) raises a `pycf3.MixedCoordinateSystemError`.

- `test_calculate_distance_multiple_delta`: call the method *calculate_distance*
  with multiple delta (`dec`, `glat` for example) raises a `pycf3.MixedCoordinateSystemError`.

**Comment by ekourkchi** [comment]
All of above test units seems to be reasonable.

I can think of a few more tests:

1) Max input distance
Throw error message if
     - NAM: distance > 38   (Mpc)
     - CF3: distance > 200   (Mpc)

2) Max velocity error
Throw error message if
    - NAM: velocity > 2400   (km/s)
    - CF3: velocity > 15000   (km/s)

**Comment by leliel12** [comment]
i already added this tests

**Comment by leliel12** [comment]
## Question

In a previous message I define these 4 tests

![image](https://user-images.githubusercontent.com/299010/100533848-b6dbda00-31e7-11eb-9a2d-c2b79d8b86f1.png)

But I think is wrong, the validation is in delta (dec, glat, sgb) not in alpha.

I checked this behavior on the cf3 page

![image](https://user-images.githubusercontent.com/299010/100533881-1639ea00-31e8-11eb-84e0-836fe5dbe7b2.png)

So I think these tests are bad. 

is this correct? @ekourkchi

**Comment by ekourkchi** [comment]
Hi,

That's right. I didn't catch the mistake the other day.

"delta" is between -90 and 90 degrees. "alpha" can basically take any number.

**Comment by leliel12** [comment]
Fix in 5c7d24c0941a98178a8de97614ea68d929c7dbca
